src/core/tsi/ssl_transport_security.cc

```diff
@@ -1450,9 +1480,19 @@ static tsi_result ssl_handshaker_next(
   if (ssl_handshaker_get_result(impl) == TSI_HANDSHAKE_IN_PROGRESS) {
     *handshaker_result = nullptr;
   } else {
-    size_t unused_bytes_size = received_bytes_size - bytes_consumed;
-    const unsigned char* unused_bytes =
-        unused_bytes_size == 0 ? nullptr : received_bytes + bytes_consumed;
+    // In TLS 1.3, the ClientFinished or ServerFinished record may have
+    // (encrypted) application data appended to the end of the record. In TLS
+    // 1.2, this is explicitly disallowed by the RFC; application data will
+    // never be appended to a handshake record.
+    unsigned char* unused_bytes = nullptr;
+    size_t unused_bytes_size = 0;
+    status = ssl_bytes_remaining(impl, &unused_bytes, &unused_bytes_size);
+    if (status != TSI_OK) return status;
+    if (unused_bytes_size > received_bytes_size) {
+      gpr_log(GPR_INFO, "More unused bytes than received bytes.");
+      gpr_free(unused_bytes);
+      return TSI_INTERNAL_ERROR;
+    }
     status = ssl_handshaker_result_create(impl, unused_bytes, unused_bytes_size,
                                           handshaker_result);
```

Comment:
It seems result_create used to take an unowned buffer and now you are passing in an owned buffer to it. Not sure whether you should release the unused_bytes after calling it.

Reply:
Great catch. I've changed `ssl_handshaker_result_create` so that it takes ownership of `unused_bytes`, and it will be freed when `ssl_handshaker_result_destroy` is called. This also saves us from doing a copy in `ssl_handshaker_result_create`.